### PR TITLE
Train functionality for DC along with unit tests

### PR DESCRIPTION
**Pull request by shantanuchhabra**

### CODE REVIEW

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 1]
```diff
@@ -25,20 +26,449 @@ using neural_net::model_backend;
...
+  gl_sframe train_data;
+  gl_sframe val_data;
+  if (variant_is<gl_sframe>(validation_data)) {
+    train_data = data;
+    val_data = variant_get_value<gl_sframe>(validation_data);
+    /* TODO: Re-evaluate the following log_and_throw.
+     *       Enabling it will trigger a unit test failure!
+     */
+    // if (val_data.empty()) {
+    //   log_and_throw("Input SFrame either has no rows or no columns. A "
```
Comment (shreyajain17) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: TODO is enough and maybe comments can be removed for the same.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 2]
```diff
@@ -25,20 +26,449 @@ using neural_net::model_backend;
...
+    log_and_throw("No neural network compute context provided");
+  }
+
+  // Report to the user what GPU(s) is being used.
+  std::vector<std::string> gpu_names = training_compute_context_->gpu_names();
+  if (gpu_names.empty()) {
+    logprogress_stream << "Using CPU to create model";
+  } else {
+    std::string gpu_names_string = gpu_names[0];
+    for (size_t i = 1; i < gpu_names.size(); ++i) {
```
Comment (shreyajain17) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: Can you use the utility function now that it is in master? https://github.com/apple/turicreate/pull/2380

File: src/ml/neural_net/tf_compute_context.hpp [hunk 3]
```diff
@@ -45,7 +45,11 @@ EXPORT class tf_compute_context: public compute_context {
       const float_array_map& config, const float_array_map& weights) override;
 
   std::unique_ptr<model_backend> create_drawing_classifier(
+<<<<<<< HEAD
```
Comment (TobyRoseman) [inline src/ml/neural_net/tf_compute_context.hpp]: This looks like a bad git merge.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 4]
```diff
@@ -25,20 +26,449 @@ using neural_net::model_backend;
...
+      std::numeric_limits<int>::max());
+  options.create_integer_option(
+      "max_iterations",
+      "Maximum number of iterations/epochs made over the data during the"
+      " training phase",
+      500,
+      1,
+      std::numeric_limits<int>::max());
+  options.create_integer_option(
+      "random_seed",
```
[the remark this answers is not in the document]
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Removed.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 1 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Done.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 2 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Done.

File: src/ml/neural_net/tf_compute_context.hpp (hunk 3 above)
Comment (shantanuchhabra) [reply src/ml/neural_net/tf_compute_context.hpp]: Thanks for it pointing out, not sure how I missed this. Fixed now.

Comment (Jarvi-Izana) [inline src/ml/neural_net/tf_compute_context.hpp]: Needs clean up.

File: src/toolkits/drawing_classifier/dc_data_iterator.hpp [hunk 5]
```diff
@@ -58,9 +58,10 @@ class data_iterator {
      */
     std::vector<std::string> class_labels;
 
-    /**
-     * Whether to traverse the data more than once.
-     */
+    /** Whether this is training data or not. */
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/dc_data_iterator.hpp]: this should be in a separate pr for dc_data_iterator. I have a dependency on it. I can hold my pr for prediction.

File: src/toolkits/drawing_classifier/drawing_classifier.hpp [hunk 6]
```diff
@@ -170,19 +186,80 @@ class EXPORT drawing_classifier: public ml_model_base {
   REGISTER_CLASS_MEMBER_FUNCTION(drawing_classifier::export_to_coreml,
                                  "filename");
 
+  REGISTER_CLASS_MEMBER_FUNCTION(drawing_classifier::init_train, "data",
+                                  "target_column_name", "feature_column_name",
+                                  "validation_data", "options");
+  register_defaults("init_train",
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.hpp]: init train is not exposed to python frontend, isn't it?

File: src/toolkits/drawing_classifier/drawing_classifier.hpp [hunk 7]
```diff
@@ -170,19 +186,80 @@ class EXPORT drawing_classifier: public ml_model_base {
...
+      gl_sframe data, bool is_train,
+      std::vector<std::string> class_labels) const;
+
+  // Factory for compute_context
+  virtual std::unique_ptr<neural_net::compute_context> create_compute_context()
+      const;
+
   // Returns the initial neural network to train
-  virtual std::unique_ptr<neural_net::model_spec> init_model() const;
+  virtual std::unique_ptr<neural_net::model_spec> init_model(
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.hpp]: this will cause merge confliction to me. let's change it back.

File: src/toolkits/drawing_classifier/drawing_classifier.hpp [hunk 8]
```diff
@@ -170,19 +186,80 @@ class EXPORT drawing_classifier: public ml_model_base {
...
+  }
 
  private:
   // Primary representation for the trained model.
   std::unique_ptr<neural_net::model_spec> nn_spec_;
-  std::unique_ptr<neural_net::model_backend> training_model_;
+
+  // Primary dependencies for training. These should be nonnull while training
+  // is in progress.
+  gl_sframe training_data_;  // TODO: Avoid storing gl_sframe AND data_iterator.
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.hpp]: why not store the data into your data_iterator?

File: test/unity/toolkits/drawing_classifier/CMakeLists.txt [hunk 9]
```diff
@@ -1,4 +1,7 @@
 project(unity_test_toolkits)
 
         make_boost_test(
-            test_dc_data_iterator.cxx REQUIRES unity_shared_for_testing)
\ No newline at end of file
+            test_dc_data_iterator.cxx REQUIRES unity_shared_for_testing)
```
Comment (Jarvi-Izana) [inline test/unity/toolkits/drawing_classifier/CMakeLists.txt]: let's indent to the left.

File: test/unity/toolkits/drawing_classifier/test_dc_data_iterator.cxx [hunk 10]
```diff
@@ -14,63 +14,13 @@
 #include <model_server/lib/image_util.hpp>
 #include <core/util/test_macros.hpp>
 
+#include "data_utils.hpp"
```
Comment (Jarvi-Izana) [inline test/unity/toolkits/drawing_classifier/test_dc_data_iterator.cxx]: again, it's better to split into 2 prs. a lot of those are related to data_iterator, which I heavily depend on both in development and test.

File: test/unity/toolkits/drawing_classifier/data_utils.hpp [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+    flex_list images(num_rows_);
+    flex_list labels(num_rows_);
+    
+    std::vector<unsigned char> buffer(IMAGE_WIDTH * IMAGE_HEIGHT * 1);
+    for (size_t i = 0; i < num_rows_; ++i) {
+
+      // Each pixel has pixel value equal to the row index (modulo 256).
+      std::fill(buffer.begin(), buffer.end(),
+                static_cast<unsigned char>(1 % 256));
```
Comment (Jarvi-Izana) [inline test/unity/toolkits/drawing_classifier/data_utils.hpp]: shouldn't it be i?

File: test/unity/toolkits/drawing_classifier/data_utils.hpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+      images[i] = flex_image(reinterpret_cast<char*>(buffer.data()),
+                             IMAGE_HEIGHT, IMAGE_WIDTH, 1, buffer.size(),
+                             IMAGE_TYPE_CURRENT_VERSION,
+                             static_cast<int>(Format::RAW_ARRAY));
```
Comment (Jarvi-Izana) [inline test/unity/toolkits/drawing_classifier/data_utils.hpp]: enum can be converted to int implicitly.

File: test/unity/toolkits/drawing_classifier/data_utils.hpp [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    params_.shuffle = false;
+    params_.class_labels = get_unique_labels();
+  }
+
+
+  /* Get the unique labels based on the generated data */
+  std::vector<std::string> get_unique_labels() const {
+    std::vector<std::string> expected_class_labels;
+    if (num_rows_ < unique_labels_.size()) {
+      std::copy(unique_labels_.begin(), unique_labels_.begin() + num_rows_,
```
Comment (Jarvi-Izana) [inline test/unity/toolkits/drawing_classifier/data_utils.hpp]: why not use insert()?

File: test/unity/toolkits/drawing_classifier/data_utils.hpp [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+              std::back_inserter(expected_class_labels));
+    } else {
+      expected_class_labels = unique_labels_;
+    }
+    return expected_class_labels;
+  }
+
+  void set_class_labels(const std::vector<std::string> &class_labels) {
```
Comment (Jarvi-Izana) [inline test/unity/toolkits/drawing_classifier/data_utils.hpp]: you shouldn't move const ref. you can use `std::vector<std::string>&&` or  `std::vector<std::string>` (which is not recommneded) in signature to indicate you want to move the argument.

File: test/unity/toolkits/drawing_classifier/data_utils.hpp [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    params_.class_labels = std::move(class_labels);
+  }
+
+  data_iterator::parameters get_iterator_params() const {
+    return params_;
+  }
+
+  gl_sframe get_data() {
```
Comment (Jarvi-Izana) [inline test/unity/toolkits/drawing_classifier/data_utils.hpp]: this can be also const since you return a copy of the gl_sframe.

File: test/unity/toolkits/drawing_classifier/data_utils.hpp (hunk 11 above)
Comment (Jarvi-Izana) [reply test/unity/toolkits/drawing_classifier/data_utils.hpp]: let's use ii instead of i. I found it's really error-prone.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 16]
```diff
@@ -25,20 +27,413 @@ using neural_net::model_backend;
...
+    // Wait until we have just one asynchronous batch outstanding. The work
+    // below should be concurrent with the neural net inference for that batch.
+    pop_until_size(1);
+
+    result result_batch;
+    result_batch.data_info = validation_data_iterator_->next_batch(batch_size);
+
+    // Submit the batch to the neural net model.
+    std::map<std::string, shared_float_array> results =
+        training_model_->predict({
```
Comment (Jarvi-Izana) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: is the predict not plugged in your test mock?

File: test/unity/toolkits/drawing_classifier/data_utils.hpp (hunk 13 above)
Comment (Jarvi-Izana) [reply test/unity/toolkits/drawing_classifier/data_utils.hpp]: Well, I think I'd better provide more explanation:

if you don't call `reserve` here to tell the container the capacity you want, it may hit its capacity and then vector will allocate new memory space with larger (x 2, usually) and copy the whole vector from previous memory space.

Usually, `back_inserter` only has local knowledge and if you want to use it, it's better to call `reserve` to guide the growth of the vector globally.

File: src/ml/neural_net/tf_compute_context.hpp (hunk 3 above)
Comment (shantanuchhabra) [reply src/ml/neural_net/tf_compute_context.hpp]: Fixed already.

File: src/toolkits/drawing_classifier/dc_data_iterator.hpp (hunk 5 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/dc_data_iterator.hpp]: Which part do you mean? Adding the `is_train` parameter?

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 16 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: Good catch, thanks for pointing out!

Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: I didn't add it because we don't expect the validation data to be non-empty. However, I agree that it's still good to mock predict though. Will add that.

File: src/toolkits/drawing_classifier/drawing_classifier.hpp [hunk 17]
```diff
@@ -6,10 +6,13 @@
 #ifndef TURI_DRAWING_CLASSIFIER_H_
 #define TURI_DRAWING_CLASSIFIER_H_
 
+#include <toolkits/drawing_classifier/dc_data_iterator.hpp>
```
Comment (shantanuchhabra) [inline src/toolkits/drawing_classifier/drawing_classifier.hpp]: Repeated import

File: src/toolkits/drawing_classifier/drawing_classifier.hpp (hunk 6 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.hpp]: Following the OD workflow, `init_train` is exposed to the Python frontend (for eventual use in checkpointing). I considered we'd want to follow that model.

File: src/toolkits/drawing_classifier/dc_data_iterator.hpp (hunk 5 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/dc_data_iterator.hpp]: yeah, that's it.

Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/dc_data_iterator.hpp]: I think the plan now is to merge this pr asap. I need `create_data_iterator`. I will switch to save and load today.

File: src/toolkits/drawing_classifier/drawing_classifier.hpp (hunk 6 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.hpp]: well, why people need a checkpoint at init_train ??? Init_train is cheap to call.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 16 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: well, let's have a common until file to aggregate all data-generators for our mock purposes.

File: src/ml/neural_net/compute_context.hpp [hunk 18]
```diff
@@ -12,6 +12,7 @@
 #include <core/export.hpp>
 #include <ml/neural_net/image_augmentation.hpp>
 #include <ml/neural_net/model_backend.hpp>
+#include <model_server/lib/variant.hpp>
```
Comment (nickjong) [inline src/ml/neural_net/compute_context.hpp]: Why was this added?

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py [hunk 19]
```diff
@@ -85,7 +84,7 @@ def __init__(self, # validation_set, # net_params,
         out = _tf.nn.xw_plus_b(fc1, weights=weights["drawing_dense1_weight"],
             biases=biases["drawing_dense1_bias"])
         out = _tf.nn.softmax(out)
-
+
```
Comment (nickjong) [inline src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: Please don't introduce trailing whitespace

File: src/toolkits/drawing_classifier/dc_data_iterator.hpp [hunk 20]
```diff
@@ -73,6 +74,9 @@ class data_iterator {
   /** Defines the output of a data_iterator. */
   struct batch {
 
+    /* Number of examples in batch */
+    size_t num_samples;
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/dc_data_iterator.hpp]: Let's give this a default value, so that a default-constructed instance of `batch` has a deterministic value

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp [hunk 21]
```diff
@@ -164,8 +179,11 @@ data_iterator::batch simple_data_iterator::next_batch(size_t batch_size) {
     }
   }
 
+  DASSERT_EQ(batch_size, batch_targets.size());
+
   // Wrap the buffers as float_array values.
   data_iterator::batch result;
+  result.num_samples = batch_size;
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/dc_data_iterator.cpp]: Will we always have `batch_size` samples? I was under the impression that DC trains by epoch... What do we do at the end of one epoch if the batch size does not evenly divide the number of training examples?

File: src/toolkits/drawing_classifier/dc_data_iterator.hpp (hunk 20 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/dc_data_iterator.hpp]: same for data_iterator::param, add default value.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 22]
```diff
@@ -25,20 +27,413 @@ using neural_net::model_backend;
...
+  else if ((variant_is<flex_string>(validation_data)) 
+      && (variant_get_value<flex_string>(validation_data)=="auto")) {
+    size_t num_rows = data.size();
+    if (num_rows >= 100) {
+      std::tie(train_data, val_data) = data.random_split(.95);
+    } else {
+      train_data = data;
+      std::cout << "The dataset has less than the minimum of 100 rows required for train-validation split. "
+                       "Continuing without validation set.\n";
+    }
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: Could we just use this implementation instead: https://github.com/apple/turicreate/blob/d06876920a172e41aa95341f0d2392d26e01d85b/src/toolkits/supervised_learning/automatic_model_creation.hpp#L28
?

File: test/unity/toolkits/drawing_classifier/CMakeLists.txt [hunk 23]
```diff
@@ -1,4 +1,7 @@
 project(unity_test_toolkits)
 
-        make_boost_test(
-            test_dc_data_iterator.cxx REQUIRES unity_shared_for_testing)
\ No newline at end of file
+make_boost_test(
+    test_dc_data_iterator.cxx REQUIRES unity_shared_for_testing)
+
+make_boost_test(
+    test_drawing_classifier.cxx REQUIRES unity_shared_for_testing)
```
Comment (nickjong) [inline test/unity/toolkits/drawing_classifier/CMakeLists.txt]: Let's add a newline here so future additions don't have to modify this line

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 16 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: After discussing offline, let's bring this in a later PR.

File: src/toolkits/drawing_classifier/drawing_classifier.hpp (hunk 8 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.hpp]: Yeah, since it's already stored in the data iterator parameters, I'll take it out.

File: test/unity/toolkits/drawing_classifier/data_utils.hpp (hunk 11 above)
Comment (shantanuchhabra) [reply test/unity/toolkits/drawing_classifier/data_utils.hpp]: Good catch, I'll fix it.

File: test/unity/toolkits/drawing_classifier/data_utils.hpp (hunk 12 above)
Comment (shantanuchhabra) [reply test/unity/toolkits/drawing_classifier/data_utils.hpp]: I'll try it, thanks.

File: test/unity/toolkits/drawing_classifier/data_utils.hpp (hunk 13 above)
Comment (shantanuchhabra) [reply test/unity/toolkits/drawing_classifier/data_utils.hpp]: Thanks, done.

File: test/unity/toolkits/drawing_classifier/data_utils.hpp (hunk 14 above)
Comment (shantanuchhabra) [reply test/unity/toolkits/drawing_classifier/data_utils.hpp]: Done.

File: src/toolkits/drawing_classifier/drawing_classifier.hpp (hunk 8 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.hpp]: Resolving per offline discussion.

File: src/ml/neural_net/compute_context.hpp (hunk 18 above)
Comment (shantanuchhabra) [reply src/ml/neural_net/compute_context.hpp]: Removed. Added by mistake when validation_data was being passed in `create_drawing_classifier`.

File: src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py (hunk 19 above)
Comment (shantanuchhabra) [reply src/python/turicreate/toolkits/drawing_classifier/_tf_drawing_classifier.py]: Sorry about that.

File: src/toolkits/drawing_classifier/drawing_classifier.hpp (hunk 6 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.hpp]: Done.

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp (hunk 21 above)
Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/dc_data_iterator.cpp]: Fixed!

File: src/toolkits/drawing_classifier/drawing_classifier.cpp [hunk 24]
```diff
@@ -69,26 +69,8 @@ std::tuple<gl_sframe, gl_sframe> drawing_classifier::init_data(
...
-    } else {
-      train_data = data;
-      std::cout << "The dataset has less than the minimum of 100 rows required for train-validation split. "
-                       "Continuing without validation set.\n";
-    }
-  } else {
-    train_data = data;
-  }
+  std::tie(train_data, val_data) = turi::supervised::create_validation_data(
+      data, validation_data);
```
Comment (nickjong) [inline src/toolkits/drawing_classifier/drawing_classifier.cpp]: Wouldn't it be more clear just to have this function be a one-liner? Just
```
return turi::supervised::create_validation_data(data, validation_data);
```

File: src/toolkits/drawing_classifier/dc_data_iterator.cpp [hunk 25]
```diff
@@ -130,25 +130,32 @@ data_iterator::batch simple_data_iterator::next_batch(size_t batch_size) {
   batch_targets.reserve(batch_size);
   batch_predictions.reserve(batch_size);
   float *next_drawing_pointer = batch_drawings.data();
+  size_t real_batch_size = 0;
+
   while (batch_targets.size() < batch_size
-      && next_row_ != range_iterator_.end()) {
+        && next_row_ != range_iterator_.end()) {
```
[the remark this answers is not in the document]
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/dc_data_iterator.cpp]: that's my bad during our peer programming. lining up with parenthesis is google style. But the double ampersand should be put at the end of the first line.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 24 above)
Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]:
```
std::tie(train_data, val_data) =
      turi::supervised::create_validation_data(data, validation_data);
```
google style.

Comment (shantanuchhabra) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: `create_validation_data` returns a `std::pair`, whereas we want to return a `std::tuple`. While `std::pair` is a special case of `std::tuple`, would directly returning `turi::supervised::create_validation_data(data, validation_data)` not throw an error?

Since we want to merge this PR asap, relevant internal tests have passed, and @Jarvi-Izana is depending on it, let's merge this now and get style changes come in later PRs.

**Comment by shantanuchhabra** [comment]
Relevant internal tests passed.

File: src/toolkits/drawing_classifier/drawing_classifier.cpp (hunk 24 above)
Comment (nickjong) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: No, it would not throw an error. `std::pair<T,U>` should be implicitly convertible to `std::tuple<T,U>`

Comment (Jarvi-Izana) [reply src/toolkits/drawing_classifier/drawing_classifier.cpp]: sounds good to me
